crate/memories/src/lib.rs

```diff
@@ -54,3 +54,22 @@ pub trait MemoryADT {
         bindings: Vec<(Self::Address, Self::Word)>,
     ) -> impl Send + std::future::Future<Output = Result<Option<Self::Word>, Self::Error>>;
 }
+
+#[cfg(feature = "batch")]
+mod batching_layer;
+
+#[cfg(feature = "batch")]
+pub use batching_layer::{MemoryBatcher, MemoryBatcherError};
+
+// Super trait for MemoryADT that allows doing write operations in batches.
+#[cfg(feature = "batch")]
+pub trait BatchingMemoryADT: MemoryADT {
+    #[allow(clippy::type_complexity)] // Refactoring this type will make the code unnecessarily more difficult to read without any actual benefit.
+    fn batch_guarded_write(
+        &self,
+        write_operations: Vec<(
+            (Self::Address, Option<Self::Word>),
+            Vec<(Self::Address, Self::Word)>,
+        )>,
+    ) -> impl Send + std::future::Future<Output = Result<Vec<Option<Self::Word>>, Self::Error>>;
```

Comment:
You define the appropriate type in `operation.rs`:
```suggestion
    fn batch_guarded_write(
        &self,
        write_operations: Vec<GuardedWriteInput<Self>>,
    ) -> impl Send + std::future::Future<Output = Result<Vec<GuardedWriteOutput<Self>>, Self::Error>>;
```

Reply:
That type was not public - I will make it public then